xunit-autolabeler-v2/ast_parser/core/analyze.py

```diff
@@ -0,0 +1,128 @@
+# Copyright 2020 Google LLC. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+from os import path
+from typing import List, Tuple
+
+from . import polyglot_drift_data as pdd, polyglot_parser, yaml_utils
+
+
+def analyze_json(
+    repo_json: str,
+    root_dir: str
+) -> Tuple[List[str], List[str], List[str], List[pdd.PolyglotDriftData]]:
+    """Perform language-agnostic AST analysis on a directory
+
+    This function processes a given directory's language-specific
+    analysis (stored in a polyglot_snippet_data.json file) into a
+    list of automatically detected snippets. It then augments the
+    automatic detection results with useful manual data (specified
+    in .drift-data.yml files). Finally, it repackages all this data
+    into a tuple containing 4 useful lists of data as shown in the
+    'returns' section.
+
+    Arguments:
+        repo_json: A path to a polyglot_snippet_data.json file
+                   generated for the specified root_dir
+        root_dir: The root directory to perform AST analysis on
+
+    Returns:
+        A tuple containing the following:
+         - A list of tags found (via grep/text search)
+           within the given directory and its subdirectories
+         - A list of tags detected (by the AST parser)
+           within the given directory and its subdirectories
+         - A list of tags that the AST parser detected,
+           but chose to ignore (due to constants or user
+           specification in .drift-data.yml files)
+         - A list of snippet objects (as typed NamedTuples)
+           detected by the AST parser in the given directory
+           and its subdirectories
+    """
+    tuple_methods = []
+    with open(repo_json, 'r') as file:
+        json_content = '\n'.join(file.readlines())
+        json_methods = json.loads(json_content)
+
+        # Normalize source_path values
+        parent_path = path.dirname(repo_json)
+        for method in json_methods:
+            method['source_path'] = (
+                path.join(parent_path, method['source_path']))
+            method['test_methods'] = (
+                [tuple(test) for test in method['test_methods']])
+
+            tuple_methods.append(
+                pdd.PolyglotDriftData(**method))
+
+    source_filepaths = set(method.source_path for method in tuple_methods)
+
+    grep_tags = set()
+    ignored_tags = set()
+
+    for source_file in source_filepaths:
```

Comment:
This function is a little long and hard to read.  Can you either break it up into sections with more comments or maybe break out some helper methods?    Like here you could either make a function for get_tags_from_files or put that in a comment showing that we moved from pulling the filepaths from the JSON and now we're scraping the files.

Helper methods would make testing easier, too.

Reply:
Agreed. 🙂

Will do.